### PR TITLE
Implement legendrank attribute in traces

### PR DESCRIPTION
**Pull request by archmoj**
Addressing #5501.

@plotly/plotly_js

### CODE REVIEW

File: src/components/legend/get_legend_data.js
Comment (alexcjohnson) [inline src/components/legend/get_legend_data.js:83]: I'm assuming `_initID` is important for pie traces? Anything else? OK, so the code below implies (and I just tested to confirm) that `legend.traceorder` flips the order of pie entries in the legend, not just the order individual traces get added to the legend.

Based on that, I'd say the correct behavior of traceorder once we add `legendrank` is to simply flip the whole thing. That's not what I thought we concluded on the call yesterday but it's what you've implemented and with the pie complication I think it's the simplest thing to explain :)

File: src/plots/attributes.js
Comment (alexcjohnson) [inline src/plots/attributes.js:51]: Let's describe what the `dflt: 1000` is for, something like:
> The default `legendrank` is 1000, so that you can use ranks less than 1000 to place certain items before all unranked items, and ranks greater than 1000 to go after all unranked items.

**Comment by alexcjohnson** [comment]
Looks good! Probably wants a mock (or just adding legendranks to some existing mock) and the description addition I commented on, then it'll be ready to go!

File: src/components/legend/get_legend_data.js
Comment (nicolaskruchten) [reply src/components/legend/get_legend_data.js:83]: I'm fine with this.

Comment (archmoj) [reply src/components/legend/get_legend_data.js:83]: > I'm assuming `_initID` is important for pie traces? Anything else? OK, so the code below implies (and I just tested to confirm) that `legend.traceorder` flips the order of pie entries in the legend, not just the order individual traces get added to the legend.

I noticed some older version of Chrome (including our image test) sort the items differently. 
So `_initID` was added in https://github.com/plotly/plotly.js/pull/5591/commits/a60a382ee0c72d5c50dc5cb36d7c4e3766c4ff42 to address that issue.

As you mentioned `pie-like` traces appear to behave differently.
In https://github.com/plotly/plotly.js/pull/5591/commits/598d60c9153baac1f03a2691554752a778d1d4a1
I was expecting that by using a higher value of `legendrank`on the first `pie` trace the slice legends come next.
However it is the opposite and it is rather confusing.

File: src/plots/attributes.js
Comment (archmoj) [reply src/plots/attributes.js]: Thanks. Please see https://github.com/plotly/plotly.js/pull/5591/commits/4ffb06c5b283b72444ab3613e4bb8b573bfac057 I also added a note to mention that ranking has impact within `legendgroup`.

**Comment by archmoj** [comment]
This PR was blocked by the old baseline system using old Chrome. 
Once #5724 merged we could simply update `legendrank` baseline and it would be ready.

File: src/components/legend/get_legend_data.js
Comment (alexcjohnson) [inline src/components/legend/get_legend_data.js:92]:
```suggestion
    var orderFn = function(a, b) {
        return (a.trace.legendrank - b.trace.legendrank) || (a._index - b._index);
    };
    for(i = 0; i < legendData.length; i++) {
        legendData[i].forEach(function(a, i) { a._index = i; });
```

Does this fix the old-chrome stability issue?

Comment (archmoj) [reply src/components/legend/get_legend_data.js]: Thanks. Applied in https://github.com/plotly/plotly.js/pull/5591/commits/d35ff08123ad49388ec565fba174a83f9f935563.

**Review by alexcjohnson — APPROVED** [review]
💃 Great!

**Comment by archmoj** [comment]
@nicolaskruchten please review and possibly give me a :star2: to merge this feature.

File: src/plots/attributes.js
Comment (nicolaskruchten) [inline src/plots/attributes.js:50]: replace "would be" with "are" please, throughout :)

Comment (archmoj) [reply src/plots/attributes.js]: Done in https://github.com/plotly/plotly.js/pull/5591/commits/a1f87b9dbe93a83e3f91da4f3599b22e9a4c166e

**Comment by nicolaskruchten** [comment]
As far as I can tell, the specific case outlined in https://github.com/plotly/plotly.js/issues/5501 is not well-handled... B's legendrank of 1 should pull its entire group into the first position:

image

**Comment by nicolaskruchten** [comment]
Here's the Pen: https://codepen.io/nicolaskruchten/pen/wvJROor

**Comment by archmoj** [comment]
@nicolaskruchten good catch. It is now fixed by https://github.com/plotly/plotly.js/pull/5591/commits/32a890c9bf513155820857ff840a229051ac2588 and tested by https://github.com/plotly/plotly.js/pull/5591/commits/a1abb19809e6c0129aeae3f89687103670ff1466.

**Comment by nicolaskruchten** [comment]
Behaviour looks good!

**Pull request merged**
